### ISSUE TITLE
testing with valgrind and/or without optimizations

### ISSUE THREAD

**Issue by pseewald** [cp2k/cp2k#994]
Is there any reason why we don't test automatically with valgrind anymore? If we accumulate memory issues due to missing testing it will get harder in the future to fix that. Also bugs related to uninitialized variables often don't show up in the testing but then cause segfaults on other systems depending on architecture/compiler.

Maybe we should also have a tester without optimizations (`-O0`) because access to uninitialized variables may be optimized away.

**Comment by oschuett** [comment]
I totally agree that we should run a Valgrind test. The [UZH test](https://dashboard.cp2k.org/archive/valgrind-sdbg/index.html) stopped working after the [libint2 upgrade](https://github.com/cp2k/cp2k/commit/20ad1a1b303df6bd121f68b2a6f3c6a576bba03e). My plan is to setup a new Valgrind test at [PC2](https://pc2.uni-paderborn.de/) - hopefully later this year.

**Comment by alazzaro** [comment]
Actually, I had a plan to install a Valgrind test on swan too (at least a SSMP version). I've started, but then I went to use valigrind4hpc and found a problem...
Likely I will not test all libraries, we can do that incrementally...

**Comment by oschuett** [comment]
One problem with Valgrind is that it instruments the binary on each startup. Since many of our tests finish very quickly, we actually spend the majority of time instrumenting. Hence, I want to implement a smarter version of the farming mode in the new `do_regtest.py`.

**Comment by oschuett** [comment]
I found a bunch of false positives, e.g. `ai_kinetic.F:173` and `atomic_kind_types.F:269` running [QS/regtest-gpw-1/Ar-2.inp](https://github.com/cp2k/cp2k/blob/master/tests/QS/regtest-gpw-1/Ar-2.inp). They disappear once I switch to `-O2`. So, apparently GCC got more aggressive with loop unrolling and we need to adopt our arch files.

**Comment by oschuett** [comment]
Turns out the [minimal test](https://dashboard.cp2k.org/archive/minimal-sdbg/index.html) already runs with `-O0` and detected the   [problem](https://storage.googleapis.com/cp2k-ci/run-cp2k-minimal-d4c88fe7_report.txt). However, because the reports were very large the dashboard refused to parse them and only turned gray.

**Comment by oschuett** [comment]
We've been running [Address Sanitizer](https://dashboard.cp2k.org/archive/asan-psmp/index.html) once per weeks for a while now.
